provider/file/file.go

```diff
@@ -18,68 +20,138 @@ var _ provider.Provider = (*Provider)(nil)
 // Provider holds configurations of the provider.
 type Provider struct {
 	provider.BaseProvider `mapstructure:",squash"`
+	Directory             string `description:"Load configuration from one or more .toml files in a directory"`
 }
 
 // Provide allows the file provider to provide configurations to traefik
 // using the given configuration channel.
 func (p *Provider) Provide(configurationChan chan<- types.ConfigMessage, pool *safe.Pool, constraints types.Constraints) error {
-	watcher, err := fsnotify.NewWatcher()
+	configuration, err := p.loadConfig()
+
 	if err != nil {
-		log.Error("Error creating file watcher", err)
 		return err
 	}
 
-	file, err := os.Open(p.Filename)
+	if p.Watch {
+		var watchDir string
+
+		if p.Directory != "" {
+			watchDir = p.Directory
+		} else {
+			watchDir = filepath.Dir(p.Filename)
```

Comment:
By adding watcher to the directory even if you just want to wach a file, you are going to watch all the directory events.
For example, if it exists 10 toml files into your directory, you are going to reload the configuration when all of these files will be modified.

That's why I suggest you to just add a watcher for the configuration file : `watchDir = p.Filename` and, maybe rename `watchDir` into `watchFile` (dir is a file ;) ) or `watchItem`...